### ISSUE TITLE
depends filter does not work with IDs

### ISSUE THREAD

**Issue by mrossinek** [GothenburgBitFactory/taskwarrior#2193]
#### To report a bug...

* What command(s) did you run?
```
% task rc.data.location:/tmp/task add test
Using alternate data.location /tmp/task
Created task 1.
Configuration override rc.data.location:/tmp/task
% task rc.data.location:/tmp/task add test2 depends:1
Using alternate data.location /tmp/task
Created task 2.
Configuration override rc.data.location:/tmp/task
% task rc.data.location:/tmp/task next depends:1
Using alternate data.location /tmp/task
Configuration override rc.data.location:/tmp/task
No matches.
```
* What did you expect to happen?
I expected that the final output results in listing the task with ID 2.
* What actually happened?
The output was empty.

This was discussed extensively earlier today over in the #taskwarrior irc channel. We found out that the `depends` filter does work but only with UUIDs. In the "rare" case where the ID matches the beginning of the UUID the above scenario might indeed appear to be working, however, in general it does not.
This contradict the man page where it is stated that:
```
depends:<id1,id2 ...>
              Declares this task to be dependent on id1 and id2.  This means that the tasks id1 and id2 should be completed before this task.  Consequently, this task will then show up on the 'blocked' report.  It accepts a comma-separated list of
              ID numbers, UUID numbers and ID ranges.  When prefixing any element of this list by '-', the specified tasks are removed from the dependency list.
```
(at least assuming that the same syntax can be used for a filter...)

* Paste the output of the ```task diag``` command.
```
% task diag
task 2.5.1
   Platform: Linux

Compiler
    Version: 8.2.1 20180831
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++11

Build Features
      Built: Nov  9 2018 23:33:36
      CMake: 3.12.4
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.5.19
 Build type: 

Configuration
       File: overwritten in example above
       Data: overwritten in example above
    Locking: Enabled
         GC: Enabled
    $EDITOR: nvim
     Server: 
         CA: 
      Trust: strict
Certificate: 
        Key: 
    Ciphers: NORMAL
      Creds: 

Hooks
     System: Enabled
   Location: /home/max/.task/hooks
     Active: on-modify.timewarrior (executable) (symlink)
   Inactive: 

Tests
      $TERM: tmux-256color (376x102)
       Dups: Scanned 763 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 763 tasks for broken references:
             No broken references found

```

**Comment by pbeckingham** [comment]
The `depends` attribute can only be used in a filter if it is done correctly, much like `tags`. It looks like a list of IDs, but it's really a concatenated string, so string matching rules apply. Use `task 2 info` and the `+BLOCKING` and `+BLOCKED` virtual tags.

**Comment by mrossinek** [comment]
Okay that is good to know! If I remember correctly, the use-case that was trying to be achieved was on the lines of: "I want a list of all tasks that depend on a given task".
Using the `+BLOCKING` and `+BLOCKED` virtual tags I cannot filter this to be relative to one task.

I suppose the best way to achieve a list like that is to use the `info` command (as you suggested) and then check the line "This task is blocking". Or is there a better way which I fail to think of right now?

If this is the way to go, I assume a small user script could be written to extract this list automatically from the `info` output and then call `task info` or `task next` on all those tasks in the list (just to document this idea somewhere).

**Comment by pbeckingham** [comment]
Reopening as a documentation problem.

Ignoring all of the above, what is it that you're trying to do here?  Not the command, but the higher level wish.  I feel there is likely an answer for you, and a FAQ entry to help.

**Comment by mrossinek** [comment]
I just remember trying to help someone figure this out over on the #taskwarrior irc channel but I don't have it in my backlog anymore.

But afaik the higher level wish was: "list all tasks that are blocked by a given task".
I can see why this is a query I would encounter when working on a project, for example when trying to make the decision which task I should work on next and then wanting some more details than what the urgency value provides me with.

I would be satisfied with what I posted before. I believe if that is posted in a FAQ that should be sufficient for documentation purposes. However, a note on this could also be added to the man page to clarify that the `depends` attribute works via string matching.

**Comment by pbeckingham** [comment]
Agreed.  Thanks.

**Comment by mooskagh** [comment]
Relevant snippet from IRC log.
```
10:27:49 crem       How can I list tasks blocked by task 34?  `task depends:34 long` returns no matches. `task long` shows that task 23 depends on 34.
10:31:55 mrossinek  crem: I currently do not have such a case to test it on but the man page suggests that task <filter> blocked or task <filter> blocking are what you are looking for
10:33:41 crem       No it doesn't work.   `task 34 blocked` and `task depends:34 blocked` return empty list, and `task 34 blocking` returns task 34, not task that is blocked by 34.
10:37:47 mrossinek  crem: Ah my bad. In my case if you just print task 34 info I get a line `This task is blocking` with the list of tasks it is blocking.
10:37:50 mrossinek  Is that what you need?
10:39:53 crem       Yes, but with all of the info of blocked tasks, last due date and priority.
10:40:09 crem       Why doesn't just filter "depends:34" work?
10:41:04 mrossinek  Well just leave out the word filter.. Simply type task depends:34
10:41:36 mrossinek  Because filter isnt a command.
10:42:49 crem       That includes my default search which is "next -BLOCKED urgency >= 0", and it doesn't show anything.   But if I type  `task depends:34 next` ("next" is default default command), it still doesn't work.
10:44:01 crem       I'm not typing "filter", I'm asking why filter "depends:34" doesn't work.  Filter is part of a query, https://taskwarrior.org/docs/filter.html
10:45:24 mrossinek  crem: I know how the filter works. If `task next depends:34 +BLOCKED` does not work I don't know how to help you.
10:45:35 mrossinek  Is your installation uptodate?
10:45:56 crem       2.5.1
10:46:12 crem       Not sure how old is that.
10:46:22 mrossinek  No thats fine.
10:46:45 mrossinek  I am using the same version. And it works fine here.
10:47:17 crem       task next depends:34 +BLOCKED` doesn't work indeed.
10:48:25 mrossinek  Okay the last guess I have then is that your ID is wrong ^^ Sorry I dont know how else to help you. What happens when you dont filter at all and simple observe the output of `task`?
10:53:34 crem       https://imgur.com/a/hbAOzIY
10:56:43 mrossinek  Since task 23 is blocking task 12 do you might have a -BLOCKING filter enabled?
10:59:31 crem       Nope. That's weird filter to have, but adding +BLOCKING doesn't help.
11:00:05 mrossinek  Okay sry then I have no idea
11:00:36 mrossinek  Did you try disabling your whole taskrc config? Just to make sure nothing in there is messing with you?
11:02:14 crem       Just tried, didn't help..
11:03:19 crem       https://gist.github.com/mooskagh/d383425367e05c3ec1d1d07039aab764 <- so this works for you?
11:06:56 mrossinek  Actually no :O
11:07:04 mrossinek  But all my other cases just worked fine?!
11:07:10 mrossinek  Okay this is indeed weird
11:07:13 mrossinek  Give me a sec
11:10:10 ddeimeke   crem: you need to use the uuid instead of the id
11:10:38 ddeimeke   crem: task 34 _uuid
11:11:03 ddeimeke   crem: task depends:$(task 34 _uuid)
11:11:11 mrossinek  Yes but why? Also, when trying to tab-complete the ID after depends: it does not suggest the newly created tasks..
11:11:16 crem       Yeah it works with uuid.. Thanks.
11:12:14 mrossinek  Well you solved it in the end.. But I wonder what might be causing this?!
11:14:42 ddeimeke   mrossinek: It is mostly because Taskwarrior uses UUIDs internally.
11:14:57 ddeimeke   mrossinek: Nevertheless this would make a good feature request.
11:15:41 mrossinek  ddeimeke: yes but earlier it did work for me with tasks that existed before but it does not work with newly created ones..
11:15:52 mrossinek  actually that is still reproducible..
11:16:29 ddeimeke   mrossinek: Earlier?
11:17:49 mrossinek  When I initially tried to help crem. I created a dummy task (which got assigned id 51) but depends on a much older task (id 4). Doing `task depends:4` then correctly listed the new task 51. However, when creating two new dummy tasks (52 and 53) where the latter depends on the former, the exact same filter does not work...
11:18:17 mrossinek  That is why it took me so long to reproduce his case since I assumed it wouldnt matter if I created two new dummy tasks rather than just one
11:19:26 mrossinek  I will take a look at this again this evening and probably open an exhaustive bug report on github. I will also try to pinpoint what makes my old task (id 4) so special. I did not try any other tasks yet so I will look into that
11:19:29 parsnip    i wonder if task 4 has a uuid that starts with 4
11:19:46 parsnip    or if 51 is the only item with a dep
11:19:59 parsnip    there may be many uuids with a 4
11:20:14 parsnip    or starting with a 4
```

**Comment by abone28** [comment]
So to filter tasks that depends on task 123, one need to run something like:
`task depends.contains:$(task 123 _uuid)`

**Comment by sdondley** [comment]
Good lord, @abone28, thank you so much! Been googling for 2 hours looking for a solution to this. Where is this "contains" property documented?

**Comment by tbabej** [comment]
@sdondley All the attribute modifiers (and their operator equivalents) are described directly in the output of `task help` :slightly_smiling_face: : 

```
Attribute modifiers make filters more precise.  Supported modifiers are:

  Modifiers         Example            Equivalent           Meaning
  ----------------  -----------------  -------------------  -------------------------
                    due:today          due = today          Fuzzy match
  not               due.not:today      due != today         Fuzzy non-match
  before, below     due.before:today   due < today          Exact date comparison
  after, above      due.after:today    due >= tomorrow      Exact date comparison
  none              project.none:      project == ''        Empty
  any               project.any:       project !== ''       Not empty
  is, equals        project.is:x       project == x         Exact match
  isnt              project.isnt:x     project !== x        Exact non-match
  has, contains     desc.has:Hello     desc ~ Hello         Pattern match
  hasnt,            desc.hasnt:Hello   desc !~ Hello        Pattern non-match
  startswith, left  desc.left:Hel      desc ~ '^Hel'        Beginning match
  endswith, right   desc.right:llo     desc ~ 'llo$'        End match
  word              desc.word:Hello    desc ~ '\bHello\b'   Boundaried word match
  noword            desc.noword:Hello  desc !~ '\bHello\b'  Boundaried word non-match
```

Alternatively `man task` touches on the topic of attribute modifiers in the "ATTRIBUTE MODIFIERS" section.

**Comment by sdondley** [comment]
Awesome, thanks! I can now see why I was getting errors trying with the new type of modifiers, too.

**Comment by smemsh** [comment]
IMO this is more than a documentation issue (ie is a good feature request).  Task IDs should be first class citizens in most or all contexts where they could be used.  Even though this field uses "string matching rules," that's an implementation detail that needn't be known to a user; enumeration of reverse dependencies is surely a common operation.  If necessary we could have e.g. an `rdepends` command that would do the required translation internally, perhaps.

See also #227 from back in 2013 in which there seemed to be concurrence that this could "override the substring search" to behave something more like an "in" list operator, however even in that issue's context it still required use of `.has`, which I would argue shouldn't be required.

Also see #1296 where a test was added that `contains/has` works, but I think it's still assumed to be a substring only, it definitely matches a lot of tasks if using eg `depends.has:1`.

**Comment by mooskagh** [comment]
Just to give more context why searching for blocked tasks is useful (for me personally).

With taskwarrior lacking proper subtasks support, I use "depends" to simulate that.
I.e. I create one "large and vague" task, and then create subtasks as dependent tasks (and then when it happens that a dependent task has to be split, I do so recursively).

So when I want to check which steps do I need to resolve the "big task", I need to search for blocked tasks (preferably, recursively).

Taskwarrior doesn't have a convenient way to do that, but I worked around it by creating a wrapper python program that does that (among other things that tw misses) for me.

----
**UPD:** I've moved the text below to the discussions: https://github.com/GothenburgBitFactory/taskwarrior/discussions/2804, it's better to comment there if someone has something to say on the general topic.

General thoughts, not really related to tw.

When I work on tasks, it happens to me more often than not, that a particular task has to be split, yet almost all task trackers that I tried, don't support that easily (the only exception that I remember is AbstractSpoon TODO list, it's great, but it's unfortunately Windows only).

So typically when there is a need for a new project, I add some task like "Implement feature XXX in YYY". When I add it, I usually don't have enough desire to think and knowledge, to tell what exactly to do.

Then when time comes, I start to add subtasks (usually 1-3 ahead, not the complete plan), something like "Research how it's done in similar projects". Then, when I reach the task, I often split it into subtasks further, and so on, until tasks are doable within a day. Also more often than not it happens, that you don't know in advance whether the task will have to be split. E.g. initially you think that you just add a library to your code and call a function, it will be a 10 minutes task, but then you end up fighting with compilation of this library for 2 days, fixing errors there and sending GitHub PRs upstream. In that case I split it into subtasks further. It's not rare for me to have "depth 5 subtask".

TW projects don't really help with that, as there's no concept of "splitting" and you kind of supposed to know most of the steps in advance.

So I wonder how do people approach that problem of having only vague tasks in the beginning and clarifying the next steps as the progress goes?
